tests/functional/src/main/java/org/mule/functional/classloading/isolation/maven/AutoDiscoverWorkspaceLocationResolver.java

```diff
@@ -0,0 +1,273 @@
+import static org.apache.commons.lang.StringUtils.isNotBlank;
+import static org.mule.runtime.core.util.Preconditions.checkNotNull;
+import static org.mule.runtime.core.util.StringMessageUtils.DEFAULT_MESSAGE_WIDTH;
+import org.mule.functional.api.classloading.isolation.WorkspaceLocationResolver;
+import org.mule.runtime.core.util.StringMessageUtils;
+
+import com.google.common.collect.Lists;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.file.FileVisitResult;
+import java.nio.file.FileVisitor;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.apache.maven.model.Model;
+import org.apache.maven.model.io.xpp3.MavenXpp3Reader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Discovers Maven projects from the rootArtifactClassesFolder folder and Maven variable
+ * {@value #MAVEN_MULTI_MODULE_PROJECT_DIRECTORY} (if present) to define the root project directory.
+ * <p/>
+ * Matches each Maven project found with the class path in order to check if it is part of the build session. When
+ * {@value #MAVEN_MULTI_MODULE_PROJECT_DIRECTORY} is not present, meaning that this not running under a Maven build session,
+ * Workspace references are resolved by filtering class path {@link URL}s if they reference to a Maven project (a target/classes
+ * or target/test-classes/ folders).
+ * <p/>
+ * If Maven surefire plugin is used to run test in Maven and the plugin has been configured with {@code forkMode=always} the
+ * following Maven system property has to be propagated on surefire configuration:
+ * 
+ * <pre>
+ *  <systemPropertyVariables>
+ *    <maven.multiModuleProjectDirectory>${maven.multiModuleProjectDirectory}</maven.multiModuleProjectDirectory>
+ *  </systemPropertyVariables>
+ * </pre>
+ * <p/>
+ *
+ * @since 4.0
+ */
+public class AutoDiscoverWorkspaceLocationResolver implements WorkspaceLocationResolver {
+
+  public static final String POM_XML_FILE = "pom.xml";
+  public static final String MAVEN_MULTI_MODULE_PROJECT_DIRECTORY = "maven.multiModuleProjectDirectory";
+
+  protected final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+  private Map<String, File> filePathByArtifactId = new HashMap<>();
+
+  /**
+   * Creates an instance of this class.
+   *
+   * @param classPath {@link URL}'s defined in class path
+   * @throws IllegalArgumentException if the rootArtifactClassesFolder doesn't point to a Maven project.
+   */
+  public AutoDiscoverWorkspaceLocationResolver(List<URL> classPath, File rootArtifactClassesFolder) {
+    checkNotNull(classPath, "classPath cannot be null");
+    checkNotNull(rootArtifactClassesFolder, "rootArtifactClassesFolder cannot be null");
+
+    File rootArtifactFolder = rootArtifactClassesFolder.getParentFile().getParentFile();
+    if (logger.isDebugEnabled()) {
```

Comment:
using logger.isDebugEnabled is redundant in this case (there are other examples of this in the PR)